presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankAccumulator.java

```diff
@@ -0,0 +1,894 @@
+        }
+        if (rightChildHeapIndex != UNKNOWN_INDEX) {
+            verify(leftChildHeapIndex != UNKNOWN_INDEX, "Left should always be inserted before right");
+            verify(rowComparisonStrategy.compare(rowId, peerGroupBuffer.getRowId(heapNodeBuffer.getPeerGroupIndex(rightChildHeapIndex))) > 0, "Max heap invariant violated");
+        }
+
+        IntegrityStats leftIntegrityStats = verifyHeapIntegrity(groupId, leftChildHeapIndex);
+        IntegrityStats rightIntegrityStats = verifyHeapIntegrity(groupId, rightChildHeapIndex);
+
+        verify(Math.abs(leftIntegrityStats.getMaxDepth() - rightIntegrityStats.getMaxDepth()) <= 1, "Heap not balanced");
+
+        return new IntegrityStats(
+                Math.max(leftIntegrityStats.getMaxDepth(), rightIntegrityStats.getMaxDepth()) + 1,
+                leftIntegrityStats.getPeerGroupCount() + rightIntegrityStats.getPeerGroupCount() + 1,
+                leftIntegrityStats.getValueCount() + rightIntegrityStats.getValueCount() + peerGroupCount);
+    }
+
+    private static class IntegrityStats
+    {
+        private final long maxDepth;
+        private final long peerGroupCount;
+        private final long valueCount;
+
+        public IntegrityStats(long maxDepth, long peerGroupCount, long valueCount)
+        {
+            this.maxDepth = maxDepth;
+            this.peerGroupCount = peerGroupCount;
+            this.valueCount = valueCount;
+        }
+
+        public long getMaxDepth()
+        {
+            return maxDepth;
+        }
+
+        public long getPeerGroupCount()
+        {
+            return peerGroupCount;
+        }
+
+        public long getValueCount()
+        {
+            return valueCount;
+        }
+    }
+
+    /**
+     * Buffer abstracting a mapping from group ID to a heap. The group ID provides the index for all operations.
+     */
+    private static class GroupIdToHeapBuffer
+    {
+        private static final long INSTANCE_SIZE = ClassLayout.parseClass(GroupIdToHeapBuffer.class).instanceSize();
+
+        /*
+         *  Memory layout:
+         *  [LONG] heapNodeIndex1,
+         *  [LONG] heapNodeIndex2,
+         *  ...
+         */
+        // Since we have a single element per group, this array is effectively indexed on group ID
+        private final LongBigArray heapIndexBuffer = new LongBigArray(UNKNOWN_INDEX);
+
+        /*
+         *  Memory layout:
+         *  [LONG] valueCount1, [LONG] heapSize1,
+         *  [LONG] valueCount2, [LONG] heapSize2,
+         *  ...
+         */
+        private final LongBigArray metricsBuffer = new LongBigArray(0);
+
+        private long totalGroups;
+
+        public void allocateGroupIfNeeded(long groupId)
+        {
+            // Group IDs generated by GroupByHash are always generated consecutively starting from 0, so observing a
+            // group ID N means groups [0, N] inclusive must exist.
+            totalGroups = Math.max(groupId + 1, totalGroups);
```

Comment:
Could we observe `N` and then `N+10` and then `N+5`?

nit: static import `max`

Reply:
Yes, we can observe N, N+10, and then N+5, but when you observe N+10, that means groups 0...N+10 exist also.